Refactoring the client and databases services

soullivaneuh opened the pull request:
Closes #7

The goal is to allow a more generic database access for reading and http/udp writing.

Related to my comment: https://github.com/Algatux/influxdb-bundle/issues/7#issuecomment-223366149

This is a WIP, lot of thing is missing. I'll update the documentation and update the PR body to explain a bit more. :wink: 
- [x] Client factory (private client services)
- [x] Database factory
- [ ] Manage multiple client/databases?
- [x] Tests. Some packages might help:
  - https://packagist.org/packages/matthiasnoback/symfony-config-test
  - https://packagist.org/packages/matthiasnoback/symfony-dependency-injection-test
- [x] Update the documentation
- [x] Deprecate not needed services/instances
- [x] Upgrade file

soullivaneuh committed:
Refactoring the client and databases services

The goal is to allow a more generic database access for reading and http/udp writing.

Algatux commented:
All ok, but please do not use the author docblok I really hate them. The autor name will be stored by git

Algatux commented:
About

> Tests. Some packages might help:
> https://packagist.org/packages/matthiasnoback/symfony-config-test
> https://packagist.org/packages/matthiasnoback/symfony-dependency-injection-test

Must think if they are really usefull

soullivaneuh commented:
@Algatux Those packages are for dev env and simplify Symfony services definition tests.

Here is an example: https://github.com/nexylan/slack-bundle/tree/master/tests/DependencyInjection

soullivaneuh commented:
> All ok, but please do not use the author docblok I really hate them. The autor name will be stored by git

Well, I not really agree but it's your project. :+1:

Algatux commented:
> @Algatux Those packages are for dev env and simplify Symfony services definition tests.
> 
> Here is an example: https://github.com/nexylan/slack-bundle/tree/master/tests/DependencyInjection

Ok! proceed you convinced me! :tada:

soullivaneuh commented:
@Algatux I'm not sure about the events. What is the utility? Is a special thing from your project or is related to influxdb-php? Will they still be useful after this PR?

Algatux commented:
> @Algatux I'm not sure about the events. What is the utility? Is a special thing from your project or is related to influxdb-php? Will they still be useful after this PR?

Events are a core feature of the bundle. In projects that use them heavily can be really usefull

soullivaneuh commented:
Ok I will take care of them so. :+1:

soullivaneuh commented:
PR updated.

I create a `ClientFactory` instead of database factory.

One HTTP and one UDP clients are created. Those service are private because the user has no need to access it (if he want it to, he can retrieve the instance from `Database`, like doctrine does for pdo conn).

For the database services, very simple: I simply use `InfluxDB\Client` as a factory with `selectDB` method: https://github.com/Algatux/influxdb-bundle/pull/10/commits/010d878bb74e5c9121ae3bb7bedb9c897b5e895c#diff-f4baa8d0ef85c42b81348c4342fc9cfbR10

soullivaneuh commented:
@Algatux I need #11 to be reviewed and merged before continue.

soullivaneuh commented:
@Algatux I started to deprecated the old services.

The problem is some other services use the deprecated ones, like this one: https://github.com/Algatux/influxdb-bundle/blob/1.0.1/src/Events/Listeners/InfluxDbEventListener.php#L34-L36

I can do it with BC kept but this will allow removing parameter typing on the constructor.

Or, I can directly all remove the deprecated classes and you push a new major directly.

What do you think? The first option will allow a smoother upgrade for users.

Algatux commented:
@Soullivaneuh need a bit more precise explanation about 

> One HTTP and one UDP clients are created. Those service are private because the user has no need to access it (if he want it to, he can retrieve the instance from Database, like doctrine does for pdo conn).

I did not understand what you have in mind :innocent:

soullivaneuh commented:
> I did not understand what you have in mind

What did you not understand precisely?

soullivaneuh commented:
For service privacy, this means:
- `algatux_influx_db.database.http` and `algatux_influx_db.database.udp` would be public and accessible  from the container
- `algatux_influx_db.client.http` and `algatux_influx_db.client.udp` because they are here only for internal stuff.

soullivaneuh commented:
> https://github.com/Algatux/influxdb-bundle/pull/10#issuecomment-223564016

@Algatux Can I have your opinion for this?

Algatux commented:
@Soullivaneuh The EventListener can use the deprecated services for now. Is there any problem that i don't see ?

soullivaneuh commented:
The constructor looks like this:

``` php
    public function __construct(
        WriterClient $httpWriter,
        WriterClient $udpWriter,
        PointsCollectionStorage $collectionStorage
    ) {
        $this->httpWriter = $httpWriter;
        $this->udpWriter = $udpWriter;
        $this->collectionStorage = $collectionStorage;
    }
```

If we want to use the new services, it will look like this:

``` php
    /**
     * @param Database                $httpDatabase
     * @param Database                $udpDatabase
     * @param PointsCollectionStorage $collectionStorage
     */
    public function __construct(
        Database $httpDatabase,
        Database $udpDatabase,
        PointsCollectionStorage $collectionStorage
    ) {
        $this->httpDatabase = $httpDatabase;
        $this->udpDatabase = $udpDatabase;
        $this->collectionStorage = $collectionStorage;
    }
```

But it's BC break. If we want to keep it BC, we must remove typehinting from parameter and do the check manually:

``` php
    /**
     * @param Database|WriterClient   $httpDatabase
     * @param Database|WriterClient   $udpDatabase
     * @param PointsCollectionStorage $collectionStorage
     */
    public function __construct(
        $httpDatabase,
        $udpDatabase,
        PointsCollectionStorage $collectionStorage
    ) {
        // Check type stuff.

        $this->httpDatabase = $httpDatabase;
        $this->udpDatabase = $udpDatabase;
        $this->collectionStorage = $collectionStorage;
    }
```

That is possible but it's a lot of workaround.

I'm used to do thing with a BC way. But your project is pretty young and not heavy downloaded.

I think we can directly remove the BC stuff and make a good `UPGRADE-2.0.md` file and directly push the next major.

What do you think?

Algatux commented:
@Soullivaneuh Leave it as it is (with the depreceated services) for now . 
We will think about refactoring the listener for a major version.

soullivaneuh commented:
> Leave it as it is (with the depreceated services) for now 

This is not a terrible idea. If we let is as is, user will always get deprecation warning they can not fix because it's internal.

Algatux commented:
Only if events are enabled in configuration. By the way I'll look at the listener soon

soullivaneuh commented:
Ok the PR is ready for me.

I updated the documentation and added an upgrade note.

Merge it if it's OK for you but please **don't tag** a new version right now.

I would like to manage multiple connections on another PR first. This one is big enough. :+1:

soullivaneuh commented:
Could you also please merge #13 before this one? It would be painless to solve conflicts with this way. :+1:

soullivaneuh commented:
@Algatux PR rebased, ready to merge! :+1: 

BTW, you can squash here. Only one commit, no broken hearts. :trollface:

soullivaneuh commented:
You can now appreciate the composer speed: https://travis-ci.org/Algatux/influxdb-bundle/jobs/135049009#L215 :wink:

Algatux commented on README.MD:
````diff
@@ -90,8 +90,11 @@ $container
 
+Or, if you prefer to defer the event:
 
+```
````
Missing ```php here

soullivaneuh replied:
Good catch! Done. :+1:
